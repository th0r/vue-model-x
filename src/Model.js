import {Watcher} from './vue-internals';
import {defineWatchersProperty, markAsStatic} from './utils';

export class Model {

  constructor() {
    // Tells Vue that it doesn't need to convert all properties of this object to reactive because
    // we explicitly mark reactive properties with `@observable` and `@computed` decorators.
    // All other properties shouldn't be reactive.
    markAsStatic(this);
  }

  update(data) {
    Object.assign(this, data);
  }

  watch(expression, callback, options) {
    if (!this.hasOwnProperty('_watchers')) {
      defineWatchersProperty(this);
    }

    if (typeof callback === 'string') {
      const methodName = callback;

      callback = this[methodName];

      if (typeof callback !== 'function') {
        throw new TypeError(`VueModel#watch: model doesn't have "${methodName}" method`);
      }
    }

    const watcher = new Watcher(this, expression, callback, options);

    if (options && options.immediate) {
      callback.call(this, watcher.value);
    }

    return () => {
      watcher.teardown();
    };
  }

}
