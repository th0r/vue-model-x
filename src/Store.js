import {Model} from './Model';

export class Store extends Model {}
