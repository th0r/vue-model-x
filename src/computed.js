import {Dep, Watcher} from './vue-internals';

const computedWatchers = Symbol('computedWatchers');
const noop = () => {};

export function computed(prototype, getterName, descriptor) {
  const getter = descriptor.get;

  if (!getter) {
    throw new TypeError(
      `${prototype.constructor.name}#${getterName} is not a getter so it can't be used as computed property`
    );
  }

  return {
    configurable: true,
    enumerable: true,
    get() {
      if (!Dep.target) {
        return getter.call(this);
      }

      if (!this.hasOwnProperty(computedWatchers)) {
        // Needed by Vue's `Watcher` constructor
        this._watchers = [];
        this[computedWatchers] = new Map();
      }

      if (!this[computedWatchers].has(prototype)) {
        this[computedWatchers].set(prototype, Object.create(null));
      }

      const watchers = this[computedWatchers].get(prototype);
      let watcher = watchers[getterName];

      if (!watcher) {
        // Lazily creating computed watcher
        watcher = watchers[getterName] = new Watcher(this, getter, noop, {lazy: true});
      }

      if (watcher.dirty) {
        watcher.evaluate();
      }

      watcher.depend();

      return watcher.value;
    }
  };
}
