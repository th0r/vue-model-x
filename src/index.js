export {observable} from './observable';
export {computed} from './computed';
export {VueModel} from './VueModel';
export {markAsStatic} from './utils';
