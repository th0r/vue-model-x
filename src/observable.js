import {Dep, defineReactive} from './vue-internals';

const observableValues = Symbol('observableValues');

export function observable(proto, name, descriptor) {
  Object.assign(descriptor, {
    configurable: true,
    writable: true
  });

  const reactiveDescriptor = {
    configurable: true,
    enumerable: true,

    get() {
      let value;

      if (this.hasOwnProperty(observableValues) && name in this[observableValues]) {
        value = this[observableValues][name];
      } else {
        value = descriptor.initializer();
        saveObservableValue(this, name, value);
      }

      if (Dep.target) {
        defineReactiveProperty(this, name, value, descriptor);
        // Should call newly created getter in order to set reactive dependency
        return this[name];
      } else {
        return value;
      }
    },

    set(value) {
      if (Dep.target) {
        defineReactiveProperty(this, name, value, descriptor);
      } else {
        if (!this.hasOwnProperty(name)) {
          // Defining reactive descriptor on the object itself in order to `Object.keys` to list it
          Object.defineProperty(this, name, reactiveDescriptor);
        }
        saveObservableValue(this, name, value);
      }
    }
  };

  return reactiveDescriptor;
}

function defineReactiveProperty(obj, name, value, originalDescriptor) {
  // Defining original property
  Object.defineProperty(obj, name, originalDescriptor);
  // Converting it to reactive
  defineReactive(obj, name, value);
}

function saveObservableValue(obj, name, value) {
  if (!obj.hasOwnProperty(observableValues)) {
    obj[observableValues] = Object.create(null);
  }

  obj[observableValues][name] = value;
}
