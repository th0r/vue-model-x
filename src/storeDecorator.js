export const storeMeta = Symbol('VueModelXStoreMeta');

export function store(meta) {
  return function storeDecorator(cls) {
    if (!meta) {
      throw new TypeError(
        /* eslint max-len: "off" */
        `VueModelX: you must provide options for "@store" decorator with at least "name" property for "${cls.name}" class`
      );
    }

    if (!meta.name) {
      throw new TypeError(`VueModelX: you must provide store name for "${cls.name}" class`);
    }

    Object.defineProperty(cls, storeMeta, {
      value: meta,
      enumerable: false
    });
  };
}
