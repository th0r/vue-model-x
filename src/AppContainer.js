import {storeMeta} from './storeDecorator';

export class AppContainer {
  constructor(state) {
    this.stores = Object.create(null);
    this.state = state || {};
  }

  getStore(StoreConstructor) {
    const meta = StoreConstructor[storeMeta];

    if (!meta) {
      throw new Error(
        `VueModelX: couldn't find store meta information for "${StoreConstructor.name}" class.\n` +
        'Did you forget to decorate it with `@store`?'
      );
    }

    let {name, deps} = meta;
    let store = this.stores[name];

    if (store) {
      return store;
    }

    store = new StoreConstructor();
    this.stores[name] = store;

    if (deps) {
      if (typeof deps === 'function') {
        deps = deps();
      }

      for (const propName of Object.keys(deps)) {
        store[propName] = this.getStore(deps[propName]);
      }
    }

    const state = this.state.hasOwnProperty(name) ? this.state[name] : undefined;

    if (store.init) {
      store.init(state);
    } else if (state) {
      throw new Error(
        `VueModelX: there is a serialized state for "${meta.name}" store but it doesn't implement "init" method`
      );
    }

    return store;
  }
}
