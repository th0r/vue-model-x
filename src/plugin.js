import {AppContainer} from './AppContainer';

export function vueModelXPlugin(Vue, opts = {}) {
  let {appContainer} = opts;
  const appContainerProp = opts.injectProperty || '$app';

  if (!appContainer) {
    appContainer = new AppContainer(opts.state);
  }

  Vue.mixin({

    beforeCreate() {
      const {stores} = this.$options;

      this[appContainerProp] = appContainer;

      if (stores) {
        for (const propName of Object.keys(stores)) {
          this[propName] = appContainer.getStore(stores[propName]);
        }
      }
    }

  });
}
