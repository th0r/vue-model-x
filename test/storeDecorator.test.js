import {store} from '../src';
import {storeMeta} from '../src/storeDecorator';

describe('@store', () => {
  it('should throw an error if decorator is called without options', () => {
    function createStore() {
      return @store()
      class Store {};
    }

    expect(createStore).toThrowErrorMatchingSnapshot();
  });

  it("should throw an error if decorator options don't provide store name", () => {
    function createStore() {
      return @store({deps: {}})
      class Store {};
    }

    expect(createStore).toThrowErrorMatchingSnapshot();
  });

  it('should attach metadata to store class', () => {
    const meta = {
      name: 'Store'
    };

    @store(meta)
    class Store {};

    expect(Store[storeMeta]).toEqual(meta);
  });
});
