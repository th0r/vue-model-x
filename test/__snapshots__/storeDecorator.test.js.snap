// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@store should throw an error if decorator is called without options 1`] = `"VueModelX: you must provide options for \\"@store\\" decorator with at least \\"name\\" property for \\"Store\\" class"`;

exports[`@store should throw an error if decorator options don't provide store name 1`] = `"VueModelX: you must provide store name for \\"Store\\" class"`;
